enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,291 @@
+
+## Motivation
+
+Today, OLM does not provide operators with the ability to block upgrades as the operator completes a critical process. Consider the following example:
+
+An operator has three versions available, v1, v2, and v3 where v2 has a long-running process that migrates resources in v1 to the new schema in v2. As of today, if operator v1 was installed with automatic updates enabled, and v2 and v3 became available at the same time, OLM would upgrade the operator from v1 to v2 to v3 without allowing v2's migration process to run. 
+
+The scenario above could place the operator in an unrecoverable state. OLM must provide operators that it manages with a method to commuincate whether or not the operator is ready to be upgraded. 
+
+### Goals
+
+- Provide opreators with the means to communicate whether or not they should be upgraded.
+- Update OLM so operators are not upgraded if they communicate that they should not be upgraded.
+- Preserve existing behavior should the operator not commuincate its upgradeable status.
+
+### Non-Goals
+
+- Support operator version rollback.
+- Provide operators with the means to cancle an upgrade that is in process.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator author, I want my operator to be able to communicate that OLM should not attempt to upgrade my operator until I remove the flag.
+
+### Implementation Details/Notes/Constraints [optional]
+
+Operators can communicate update readiness by exposing an `upgradeReadiness` endpoint. This endpoint can communicate readiness by:
+
+- printing 0, indicating that the operator cannot be upgraded at this time.
+- printing 1, indicating that the operator is ready to be upgraded.
+
+This endpoint will be exposed via a [Service resource](https://kubernetes.io/docs/concepts/services-networking/service/#service-resource) that OLM will probe for update readiness prior to upgrading the operator to the next version.
+
+This feature must be supported by both CSV and CSVless bundles.
+
+#### CSV Bundles
+
+OLM's [ClusterServiceVersion](https://github.com/operator-framework/api/blob/197407cd70e8ddfef85d21216085ed52fbb4bb2d/pkg/operators/v1alpha1/clusterserviceversion_types.go#L500) type will be updated to include a upgradeableService for each [StrategyDeploymentSpec](https://github.com/operator-framework/api/blob/197407cd70e8ddfef85d21216085ed52fbb4bb2d/pkg/operators/v1alpha1/clusterserviceversion_types.go#L63).
+
+There are two options for defining a Upgradeable Readiness Service in the CSV.
+
+##### Approach 1: Define the service
```

Comment:
Since CSV's are going away anyway, wouldn't it be best to only support approach 2?

Reply:
Although OLM plans to deprecate the CSV, I do not believe that we have reached the point where OLM will stop supporting new features in the CSV.

CC @njhale @ecordell